### PR TITLE
Index Argo application from unstructured data

### PR DESCRIPTION
**Pull request by KevinFCormier**
**Related Issue:**  open-cluster-management/backlog#9514

### CODE REVIEW

**Comment by KevinFCormier** [comment]
@jlpadilla Can I get your initial thoughts on this? I still need to add some testing and perhaps refactor a bit, but I wanted to see what you think of this approach as I don't have much of experience with golang.  I initially tried importing the type from ArgoCD (see https://github.com/open-cluster-management/search-collector/pull/103), but to get that compiling, I needed to update a lot of packages to a newer version and make a few minor code edits. But then I hit this runtime error, which seemed to be something in the type from ArgoCD that I probably can't easily solve:
```
E0309 21:13:15.715463       1 transformer.go:446] Error in transformer routine: reflect: reflect.Value.Set using value obtained using unexported field
E0309 21:13:15.715575       1 transformer.go:447] goroutine 24 [running]:
runtime/debug.Stack(0x2d2f020, 0x2, 0x1cb6d71)
	/home/travis/.gimme/versions/go1.15.8.linux.amd64/src/runtime/debug/stack.go:24 +0x9f
github.com/open-cluster-management/search-collector/pkg/transforms.handleRoutineExit(0xc00041c360, 0xc00041c3c0)
	/home/travis/gopath/src/github.com/open-cluster-management/search-collector/pkg/transforms/transformer.go:447 +0xaa
panic(0x19f1560, 0xc00062d750)
	/home/travis/.gimme/versions/go1.15.8.linux.amd64/src/runtime/panic.go:969 +0x1b9
reflect.flag.mustBeAssignableSlow(0x1a1)
	/home/travis/.gimme/versions/go1.15.8.linux.amd64/src/reflect/value.go:257 +0x1b9
reflect.flag.mustBeAssignable(...)
	/home/travis/.gimme/versions/go1.15.8.linux.amd64/src/reflect/value.go:247
reflect.Value.Set(0x19ad6a0, 0xc0004809b0, 0x1a1, 0x19ad6a0, 0xc000e96a70, 0x81)
	/home/travis/.gimme/versions/go1.15.8.linux.amd64/src/reflect/value.go:1558 +0x3b
k8s.io/apimachinery/pkg/runtime.structFromUnstructured(0x1a54d80, 0xc0023ede90, 0x15, 0x1b5b3a0, 0xc000480980, 0x199, 0x7fc711da6108, 0xc00062d710)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:386 +0x30f
k8s.io/apimachinery/pkg/runtime.fromUnstructured(0x1a54d80, 0xc0023ede90, 0x15, 0x1b5b3a0, 0xc000480980, 0x199, 0x1a317a0, 0xc00062d710)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:219 +0x3f1
k8s.io/apimachinery/pkg/runtime.structFromUnstructured(0x1a54d80, 0xc0023ede60, 0x15, 0x1b9e600, 0xc000480918, 0x199, 0x7fc711da6108, 0xc00062d6b0)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:382 +0x34b
k8s.io/apimachinery/pkg/runtime.fromUnstructured(0x1a54d80, 0xc0023ede60, 0x15, 0x1b9e600, 0xc000480918, 0x199, 0x1a317a0, 0xc00062d6b0)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:219 +0x3f1
k8s.io/apimachinery/pkg/runtime.structFromUnstructured(0x1a54d80, 0xc0023eced0, 0x15, 0x1b67a40, 0xc000480800, 0x199, 0xc001a8bc60, 0xc001a8ba88)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:382 +0x34b
k8s.io/apimachinery/pkg/runtime.fromUnstructured(0x1a54d80, 0xc0023eced0, 0x15, 0x1b67a40, 0xc000480800, 0x199, 0x203000, 0xc000403680)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:219 +0x3f1
k8s.io/apimachinery/pkg/runtime.(*unstructuredConverter).FromUnstructured(0x2d236b0, 0xc0023eced0, 0x1c76580, 0xc000480800, 0xc00050f410, 0xb)
	/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/runtime/converter.go:126 +0x2ba
github.com/open-cluster-management/search-collector/pkg/transforms.TransformRoutine(0xc00041c360, 0xc00041c3c0)
	/home/travis/gopath/src/github.com/open-cluster-management/search-collector/pkg/transforms/transformer.go:198 +0x3fd
created by github.com/open-cluster-management/search-collector/pkg/transforms.NewTransformer
	/home/travis/gopath/src/github.com/open-cluster-management/search-collector/pkg/transforms/transformer.go:151 +0xd1
```

**Comment by KevinFCormier** [comment]
/cc @jlpadilla
Test coverage is added and this is ready for review.

File: pkg/transforms/argoapplication_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
+IBM Confidential
+OCO Source Materials
+(C) Copyright IBM Corporation 2019 All Rights Reserved
+The source code for this program is not published or otherwise divested of its trade secrets,
+irrespective of what has been deposited with the U.S. Copyright Office.
+Copyright (c) 2020 Red Hat, Inc.
+*/
```
Comment (jlpadilla) [inline pkg/transforms/argoapplication_test.go:8]: Don't need these in a new file.
```suggestion
```

File: pkg/transforms/argoapplication_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+package transforms
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestTransformArgoApplication(t *testing.T) {
+	u := &unstructured.Unstructured{
```
Comment (jlpadilla) [inline pkg/transforms/argoapplication_test.go:20]: This works, but can we read from a json file for consistency with the other resources?

**Comment by jlpadilla** [comment]
> @jlpadilla Can I get your initial thoughts on this? I still need to add some testing and perhaps refactor a bit, but I wanted to see what you think of this approach as I don't have much of experience with golang. I initially tried importing the type from ArgoCD (see #103), but to get that compiling, I needed to update a lot of packages to a newer version and make a few minor code edits. But then I hit this runtime error, which seemed to be something in the type from ArgoCD that I probably can't easily solve:

@KevinFCormier Sorry I missed this question. The correct way in go is to import the library, but it is a pain to keep up with the dependencies. Another option is to define the structure of the Argo object.  See the PolicyReport transform for an example.
https://github.com/open-cluster-management/search-collector/blob/main/pkg/transforms/policyreport.go#L12-L34

(Comming from javascript I also have trouble with unstructured data in go.)

**Comment by jlpadilla** [comment]
BTW, this implementation also works, the only worry is consistency.

File: pkg/transforms/argoapplication_test.go (hunk 1 above)
Comment (KevinFCormier) [reply pkg/transforms/argoapplication_test.go]: Copy paste error. Thanks.

**Comment by KevinFCormier** [comment]
Closing in favour of #114

**Pull request closed without merging**